### ISSUE TITLE
Slow performance over large unions with ForwardRefs

### ISSUE THREAD

**Issue by kyle-query** [microsoft/pyright#6594]
**Describe the bug**

After changing a type definition from a "base class" to a Union over the previous subclasses, pyright now takes over 3 minutes to type check a large project, compared to 45 seconds previously.

One file in particular takes over 2 minutes to check, and I was able to narrow it further to two functions. While I cannot share the code, I tried to create a small example that might demonstrate the issue. In the real function there are around 20 calls to `__and__` in each method. Replacing them with `pass` reduces the time to check the real file to ~5 seconds.

Here's an approximation of what's happening,

```python
from abc import ABC
from typing import Union, TypeVar, Generic, cast, Any

T = TypeVar("T")
U = TypeVar("U")


FooU = Union["F0[T, U]", "F1[T, U]", "F2[T, U]", "F3[T, U]", "F4[T, U]", "F5[T, U]", "F6[T, U]", "F7[T, U]", "F8[T, U]", "F9[T, U]", "G0[T]", "G1[T]", "G2[T]", "H0", "H1"]


class _Foo(Generic[T, U], ABC):
    def __and__(self, other: "FooU[T, U]") -> "FooU[T, U]":
        return cast("FooU[T, U]", self)

class _Bar(_Foo[T, U]):
    pass

class F0(_Bar[T, U]):
    pass

class F1(_Bar[T, U]):
    pass

class F2(_Bar[T, U]):
    pass

class F3(_Bar[T, U]):
    pass

class F4(_Bar[T, U]):
    pass

class F5(_Bar[T, U]):
    pass

class F6(_Bar[T, U]):
    pass

class F7(_Bar[T, U]):
    pass

class F8(_Bar[T, U]):
    pass

class F9(_Bar[T, U]):
    pass

class G0(_Foo[T, Any]):
    pass

class G1(_Foo[T, Any]):
    pass

class G2(_Foo[T, Any]):
    pass

class H0(_Foo[Any, Any]):
    pass

class H1(_Foo[Any, Any]):
    pass

FooV = Union[F0[T, U], F1[T, U], F2[T, U], F3[T, U], F4[T, U], F5[T, U], F6[T, U], F7[T, U], F8[T, U], F9[T, U], G0[T], G1[T], G2[T], H0, H1]

def test(x: FooU[T, U], y: FooU[T, U]):
    print(x & y)
    print(x & y)
    print(x & y)
    print(x & y)
    print(x & y)
    print(x & y)
    print(x & y)
    print(x & y)
    print(x & y)
    print(x & y)
    print(x & y)
    print(x & y)
```

On my machine, using strict mode takes ~1.6 seconds on the above example.

```
$ poetry run pyright --stats test.py
WARNING: there is a new pyright version available (v1.1.335 -> v1.1.338).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`

Loading pyproject.toml file at /Users/kyle/tmp/pyproject.toml
Found 1 source file
pyright 1.1.335
0 errors, 0 warnings, 0 informations
Completed in 1.645sec

Analysis stats
Total files parsed and bound: 18
Total files checked: 1

Timing stats
Find Source Files:    0sec
Read Source Files:    0.01sec
Tokenize:             0.04sec
Parse:                0.05sec
Resolve Imports:      0.04sec
Bind:                 0.04sec
Check:                1.13sec
Detect Cycles:        0sec
```

These factors seem to affect how much time is taken:

* Increasing/decreasing the number of types in the Union
* Increasing/decreasing the number of type parameters to _Foo
* Increasing/decreasing the number of calls to `+`.

Changing in `test`'s argument types  from `FooU[T, U]` to `FooV[T, U]` drops the time by ~0.4 seconds. Changing them to `FooV[int, int]` drops it another ~0.2 seconds. Replacing `"FooU[T, U]"` in the definition of `__and__` with `"FooV[T, U]"` drops it another ~0.1 seconds. It seems like resolving ForwardRefs is contributing quite a bit, as the these changes cumulatively reduce the time from ~1.6 seconds to ~0.8 seconds.

The time to lookup the type of `+` when the receiver is known to be `FooU[T, U]` also seems to be a big contributor to the time required. I'm guessing pyright looks up this method on each type in the Union. Making no changes to the original example except replacing any reference to `FooU[K, V]` with `_Foo[K, V]` reduces the time to 0.6 seconds.

**VS Code extension or command-line**

This can be reproduced with pyright 1.1.335

**Comment by erictraut** [comment]
When dealing with binary operations (such as `&`) and operands whose types are unions, type analysis time will be proportional to `n x m` where `n` is the number of subtypes in the left operand's union and `m` is the number of subtypes in the right operand's union. Your union contains 15 elements, so evaluating `x & y` effectively involves evaluating 225 expressions.

Note that `&` is more complicated than a simple call to `__and__`. If the `__and__` fails to type check, then pyright needs to evaluate the `__rand__` on the RH operand. (This doesn't appear to apply in your code, but if it did, we'd need to multiply by 2x to account for this.)

If there are type variables that need to be solved (as opposed to concrete types), then each call must be evaluated `x + 1` times where `x` is the number of independent type variables that need to be solved. In your case, there are two type variables (`T` and `U`), so this adds a factor of 3 giving us 675. If there are 20 of these per function, that means 13,500 expression evaluations per function. As you can see, these numbers grow quickly when you start multiplying them.

What do you mean by "calls to `+`"? Did you mean "calls to `__and__`"?

I'm not seeing any appreciable difference between `FooU` and `FooV`. I wouldn't expert there to be. Most of pyright's logic is indifferent to whether references are forward or backward.

I'll investigate more deeply to see if there's a smoking gun, but I'm guessing that you're simply experiencing the consequences of a combinatoric explosion. This can happen in type checkers when dealing with complex types — especially unions with many subtypes.

**Comment by erictraut** [comment]
Does the `__and__` method really return any of the subtypes in the union? Or does it always return the same type as `self`? If I change its definition to the following, performance is significantly improved.

```python
class _Foo(Generic[T, U], ABC):
    def __and__(self, other: "_Foo[T, U]") -> Self:
        return self
```

**Comment by kyle-query** [comment]
> What do you mean by "calls to +"? Did you mean "calls to __and__"?

Opps, I meant `__and__` yes.

> Your union contains 15 elements, so evaluating x & y effectively involves evaluating 225 expressions.

That makes sense. In the real code I think there are 25 elements in the union, so I can see how this could be quite expensive.

In some issues you've mentioned pyright using a cache for various things. I wonder if that could be applied here, since evaluating these same types over and over is coming at a high cost?

> Does the __and__ method really return any of the subtypes in the union? Or does it always return the same type as self? If I change its definition to the following, performance is significantly improved.

Hmm that's interesting. In the real code, `__and__` can only return 3 of the subtypes in the Union.

Another interesting thing is in the real code, replacing calls to `&` with a call to this function:

```python
def _and(x: FooU[int, int], y: FooU[int, int]) -> FooU[int, int]:
    return x & y
```

Reduces the time to typecheck from ~110 seconds to ~10 seconds. I should note in the real code, the two type variables are already instantiated to concrete types at the site where `&` is called.

**Comment by erictraut** [comment]
> replacing calls to & with a call to this function ... reduces the time...

Yeah, that makes sense. A call to a simple function represents a single call, and it can be done with one evaluation. A method must be bound to the object it is called through. This needs to be done separately for each of the types in the LH operand. If that type is a union of 25 different types, it means the binding must be done 25 times, and this results in 25 calls that must be independently evaluated.

**Comment by erictraut** [comment]
I've spent some additional time on this. I found a few small optimizations on some hot paths, but I wasn't able to remove the primary sources of the combinatoric explosion that results from the structure of your code.

In object-oriented programming, a base class should not have knowledge of its subclasses. If such a class operates on subclasses, the proper way to model that is with generics. The `_Foo.__and__` method in your code sample above violates this basic programming principle.

Here are two approaches that eliminate this problem — and also eliminate the performance issue that you're seeing. Since you haven't shown the real implementation of `__and__` in your example above, I'm not sure which approach more closely matches the actual implementation.

1. Accept any `_Foo[T, U]` as an input parameter and return `Self` as a return type.
```python
class _Foo(Generic[T, U], ABC):
    def __and__(self, other: "_Foo[T, U]") -> Self:
        return self
```

2. Use a bound TypeVar for the input parameter and return this same type as a return type.
```python
TFoo = TypeVar("TFoo", bound="_Foo[Any, Any]")

class _Foo(Generic[T, U], ABC):
    def __and__(self, other: TFoo) -> TFoo:
        return cast(TFoo, self)
```

Using either of these approaches will decrease the type checking time by several orders of magnitude.

I don't plan to spend additional time on this, so I'm going to close the issue.
